### PR TITLE
Clarify events.

### PR DESCRIPTION
**Pull request by chriseth**
Fixes https://github.com/ethereum/solidity/issues/4167

### CODE REVIEW

File: docs/contracts.rst [hunk 1]
```diff
@@ -723,22 +723,17 @@ the contract can only see the last 256 block hashes.
...
-to be searched for: It is possible to filter for specific values of
-indexed arguments in the user interface.
-
-If arrays (including ``string`` and ``bytes``) are used as indexed arguments, the
-Keccak-256 hash of it is stored as topic instead.
-
-The hash of the signature of the event is one of the topics except if you
+to be stored in a special data structure as so-called "topics", which allows them to be searched for,
+for example when filtering a sequence of blocks for certain events. Events can always
+be filtered by the address of the cantract that emitted the event. Also,
```
Comment (axic) [inline docs/contracts.rst]: cantract -> contract

File: docs/contracts.rst [hunk 2]
```diff
@@ -723,22 +723,17 @@ the contract can only see the last 256 block hashes.
...
 Up to three parameters can
 receive the attribute ``indexed`` which will cause the respective arguments
-to be searched for: It is possible to filter for specific values of
-indexed arguments in the user interface.
-
-If arrays (including ``string`` and ``bytes``) are used as indexed arguments, the
-Keccak-256 hash of it is stored as topic instead.
-
-The hash of the signature of the event is one of the topics except if you
+to be stored in a special data structure as so-called "topics", which allows them to be searched for,
```
Comment (axic) [inline docs/contracts.rst:726]: Should mention the size of each topic (e.g. 256bit?). Perhaps a link to a description of these could be useful, if there's anything outside the YP :)

File: docs/contracts.rst [hunk 3]
```diff
@@ -723,22 +723,17 @@ the contract can only see the last 256 block hashes.
...
+for example when filtering a sequence of blocks for certain events. Events can always
+be filtered by the address of the cantract that emitted the event. Also,
+the hash of the signature of the event is one of the topics except if you
 declared the event with ``anonymous`` specifier. This means that it is
 not possible to filter for specific anonymous events by name.
 
-All non-indexed arguments will be stored in the data part of the log.
+If arrays (including ``string`` and ``bytes``) are used as indexed arguments, the
```
Comment (axic) [inline docs/contracts.rst:733]: Shouldn't this also mention that "any type larger than 256 bits" is handled like this?

File: docs/contracts.rst [hunk 4]
```diff
@@ -723,22 +723,17 @@ the contract can only see the last 256 block hashes.
...
+Keccak-256 hash of it is stored as topic instead.
 
-.. note::
-    Indexed arguments will not be stored themselves.  You can only
-    search for the values, but it is impossible to retrieve the
-    values themselves.
+All non-indexed arguments will be ABI-encoded into the data part of the log.
```
Comment (axic) [inline docs/contracts.rst]: "ABI-encoded" should be a link to the ABI chapter.

File: docs/contracts.rst (hunk 2 above)
Comment (chriseth) [reply docs/contracts.rst:726]: I don't think the size is relevant. We also do not explain how "topics" are stored, so why should we care about the size? I will explain below why non-word values are hashed.

File: docs/contracts.rst (hunk 3 above)
Comment (chriseth) [reply docs/contracts.rst:733]: structs are not supported, so this is fine. Ah, let me open an issue for structs actually.

**Comment by chriseth** [comment]
Updated.

**Review by axic — APPROVED** [review]

**Comment by fulldecent** [comment]
Thank you, LGMM

**Pull request merged**